blockchain/validate.go

```diff
@@ -2400,29 +2796,46 @@ func CheckTransactionInputs(subsidyCache *standalone.SubsidyCache, tx *dcrutil.T
 // input and output scripts in the provided transaction.  This uses the
 // quicker, but imprecise, signature operation counting mechanism from
 // txscript.
-func CountSigOps(tx *dcrutil.Tx, isCoinBaseTx bool, isSSGen bool) int {
+func CountSigOps(tx *dcrutil.Tx, isCoinBaseTx bool, isSSGen bool, isTreasuryEnabled bool) int {
 	msgTx := tx.MsgTx()
 
-	// Accumulate the number of signature operations in all transaction
-	// inputs.
+	var isTreasuryBase, isTSpend, isTAdd bool
+	if isTreasuryEnabled {
+		isTreasuryBase = stake.IsTreasuryBase(msgTx)
+		isTSpend = stake.IsTSpend(msgTx)
+		isTAdd = stake.IsTAdd(msgTx)
+	}
+
 	totalSigOps := 0
-	for i, txIn := range msgTx.TxIn {
-		// Skip coinbase inputs.
-		if isCoinBaseTx {
-			continue
-		}
-		// Skip stakebase inputs.
-		if isSSGen && i == 0 {
-			continue
+
+	// isTreasuryBase has no sigops in input and output so just return.
+	if isTreasuryBase {
+		return totalSigOps
+	}
+
+	if !(isCoinBaseTx || isTSpend) {
+		// Accumulate the number of signature operations in all
+		// transaction inputs.
+		for i, txIn := range msgTx.TxIn {
+			// Skip stakebase inputs.
+			if isSSGen && i == 0 {
+				continue
+			}
+
+			numSigOps := txscript.GetSigOpCount(txIn.SignatureScript)
+			totalSigOps += numSigOps
 		}
+	}
 
-		numSigOps := txscript.GetSigOpCount(txIn.SignatureScript)
-		totalSigOps += numSigOps
+	// TAdds have not outputs so just return.
```

Comment:
TAdds may have a (sstx-)change output that is (currently) enforced to be a P2PKH or P2SH so they may have 1 sigop.

Reply:
You are right and @davecgh and I went through this in detail.